lib/ts/nextjs/middleware.ts

```diff
@@ -0,0 +1,323 @@
+
+import {
+    REDIRECT_ATTEMPT_MAX_COUNT,
+    REFRESH_TOKEN_COOKIE_NAME,
+    REFRESH_TOKEN_HEADER_NAME,
+    REDIRECT_PATH_PARAM_NAME,
+    REDIRECT_ATTEMPT_COUNT_COOKIE_NAME,
+    ACCESS_TOKEN_COOKIE_NAME,
+    ACCESS_TOKEN_HEADER_NAME,
+    FRONT_TOKEN_HEADER_NAME,
+    ANTI_CSRF_TOKEN_HEADER_NAME,
+    FORCE_LOGOUT_PATH_PARAM_NAME,
+    FRONT_TOKEN_COOKIE_NAME,
+    ANTI_CSRF_TOKEN_COOKIE_NAME,
+    CURRENT_PATH_COOKIE_NAME,
+    SESSION_REFRESH_API_PATH,
+} from "./constants";
+
+import type { ApiRequestMiddleware, SuperTokensNextjsConfig } from "./types";
+
+let AppInfo: SuperTokensNextjsConfig["appInfo"];
+const DEFAULT_API_BASE_PATH = "/api/auth";
+
+export function superTokensMiddleware(
+    config: SuperTokensNextjsConfig,
+    apiRequestMiddleware?: ApiRequestMiddleware
+): (request: Request) => Promise<Response | void> {
+    const usesTheNextjsApiAsTheAuthenticationServer = compareUrlHost(
+        config.appInfo.apiDomain,
+        config.appInfo.websiteDomain
+    );
+
+    return async (request: Request) => {
+        const requestUrl = new URL(request.url);
+        if (requestUrl.pathname.startsWith(SESSION_REFRESH_API_PATH) && request.method === "GET") {
+            return refreshSession(config, request);
+        }
+
+        if (requestUrl.pathname.startsWith("/api") && usesTheNextjsApiAsTheAuthenticationServer) {
+            if (requestUrl.pathname.startsWith(config.appInfo.apiBasePath || DEFAULT_API_BASE_PATH)) {
+                // this hits our pages/api/auth/* endpoints
+                return next();
+            }
+
+            if (apiRequestMiddleware) {
+                return apiRequestMiddleware(request);
+            }
+        }
+
+        if (
+            requestUrl.pathname.startsWith("/auth") &&
+            requestUrl.searchParams.get(FORCE_LOGOUT_PATH_PARAM_NAME) === "true"
+        ) {
+            return revokeSession(config, request);
+        }
+
+        // Save the current path so that we can use it during SSR
+        // Used to redirect the user to the correct path after login/refresh
+        // https://github.com/vercel/next.js/issues/43704#issuecomment-2090798307
+        // TL;DR: You can not access pathname in SSR and requests that originate from redirect()
+        const response = next();
+        response.headers.append(
+            "set-cookie",
+            `${CURRENT_PATH_COOKIE_NAME}=${requestUrl.pathname}; Path=/; HttpOnly; SameSite=Strict`
+        );
+        return response;
+    };
+}
+
+export async function refreshSession(config: SuperTokensNextjsConfig, request: Request): Promise<Response> {
+    AppInfo = config.appInfo;
+    if (config.enableDebugLogs) {
+        enableLogging();
+    }
+
+    // Cancel the refresh cycle if an unforseen state is encountered
+    const redirectAttemptNumber = getRedirectAttemptNumber(request);
+    if (redirectAttemptNumber > REDIRECT_ATTEMPT_MAX_COUNT) {
+        return redirectToAuthPage(request);
+    }
```

Comment:
This is so good... we may want to add this in backend SDKs in general.

Reply:
Renamed this into `confirmAuthenticationAndCallServerAction` since `authenticateServerAction` might lead users to wrong conclusions. It just checks the authentication state (and retries if required). We can't do the flow where we call the server action first and check for a 401 since serve actions can't return 401 responses.

If they want to access the authentication session object inside the server action they need to call the other function, 'getServerActionSession`.